## <center> Q&A Activity Week - 11

### Reading the dataset to python environment

In [1]:
import pandas as pd
Student = pd.read_csv('student-data.csv')

In [2]:
Student # to get an idea about the structure of the data set

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... internet romantic  famrel  freetime  goout Dalc Walc health absences  \
0    ...       no       no       4         3      4    1    1      3        6   
1    ...      yes       no       5         3      3    1    1      3        4   
2    ...      yes       no       4         3      2    2    3      3       10   
3    ...      yes      yes       3         2      2    1    1      5        2   
4    ...       no       no       4         3      2    1    2      5        4   
..   ...      ...      ...     ...       ...    ...  ...  ...    ...      ...   
390  ...       no       no       5         5      4    4    5      4       11   
391  ...      yes       no       2         4      5    3    4      2        3   
392  ...       no       no       5         5      3    3    3      3        3   
393  ...      yes       no       4         4      1    3    4      5        0   
394  ...      yes       no       3         2      3    3    3      5        5   

    passed  
0       no  
1       no  
2      yes  
3      yes  
4      yes  
..     ...  
390     no  
391    yes  
392     no  
393    yes  
394     no  

[395 rows x 31 columns]

In [3]:
Student.info()       # to get an overall view about data set

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 31 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

Dataset consist of  395 rows and 31 columns 

No null values are present in the data set

Exploring the dataset to determine how many students we have information on, and learn about the graduation rate among these students.

In [4]:
students = len(Student.index)
passed = len(Student[(Student.passed == 'yes')])
failed = len(Student[(Student.passed == 'no')])
grad_rate = float(passed)/float(students)*100.0

# Print the results
print ("Total number of students: {}".format(students))
print ("Number of students who passed: {}".format(passed))
print ("Number of students who failed: {}".format(failed))
print ("Graduation rate of the class: {:.2f}%".format(grad_rate))

Total number of students: 395
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


Since the current graduation rate is 67.09% and the intended rate after ten years is 95%,timely intervention is necessary.

#### Preprocessing the dataset

In [5]:
#before preprocessing identifying columns that require encoding 
for i in Student.columns:
    print(i,Student[i].unique(),Student[i].dtype)

school ['GP' 'MS'] object
sex ['F' 'M'] object
age [18 17 15 16 19 22 20 21] int64
address ['U' 'R'] object
famsize ['GT3' 'LE3'] object
Pstatus ['A' 'T'] object
Medu [4 1 3 2 0] int64
Fedu [4 1 2 3 0] int64
Mjob ['at_home' 'health' 'other' 'services' 'teacher'] object
Fjob ['teacher' 'other' 'services' 'health' 'at_home'] object
reason ['course' 'other' 'home' 'reputation'] object
guardian ['mother' 'father' 'other'] object
traveltime [2 1 3 4] int64
studytime [2 3 1 4] int64
failures [0 3 2 1] int64
schoolsup ['yes' 'no'] object
famsup ['no' 'yes'] object
paid ['no' 'yes'] object
activities ['no' 'yes'] object
nursery ['yes' 'no'] object
higher ['yes' 'no'] object
internet ['no' 'yes'] object
romantic ['no' 'yes'] object
famrel [4 5 3 1 2] int64
freetime [3 2 4 1 5] int64
goout [4 3 2 1 5] int64
Dalc [1 2 5 3 4] int64
Walc [1 3 2 4 5] int64
health [3 5 1 2 4] int64
absences [ 6  4 10  2  0 16 14  7  8 25 12 54 18 26 20 56 24 28  5 13 15 22  3 21
  1 75 30 19  9 11 38 40 23 17] int64


In [6]:
# Converting all categorical columns to numerical columns 

In [7]:
Student.replace(['yes', 'no'], [1, 0],inplace =True) 

In [8]:
for x in Student.columns:
        if Student[x].dtype==object:
            dummies = pd.get_dummies(Student[x], prefix=x, dummy_na=False)
            Student = Student.drop(x,axis=1)
            Student = pd.concat([Student, dummies], axis=1)

In [9]:
Student.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 49 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   age                395 non-null    int64
 1   Medu               395 non-null    int64
 2   Fedu               395 non-null    int64
 3   traveltime         395 non-null    int64
 4   studytime          395 non-null    int64
 5   failures           395 non-null    int64
 6   schoolsup          395 non-null    int64
 7   famsup             395 non-null    int64
 8   paid               395 non-null    int64
 9   activities         395 non-null    int64
 10  nursery            395 non-null    int64
 11  higher             395 non-null    int64
 12  internet           395 non-null    int64
 13  romantic           395 non-null    int64
 14  famrel             395 non-null    int64
 15  freetime           395 non-null    int64
 16  goout              395 non-null    int64
 17  Dalc            

#### Feature Engineering

In [10]:
# Adding up mothers education and fathers education to Parents education,denoted by Pedu
Student['Pedu'] = Student['Medu'] + Student['Fedu']
Student.drop(['Medu','Fedu'],axis=1)

age  traveltime  studytime  failures  schoolsup  famsup  paid  \
0     18           2          2         0          1       0     0   
1     17           1          2         0          0       1     0   
2     15           1          2         3          1       0     1   
3     15           1          3         0          0       1     1   
4     16           1          2         0          0       1     1   
..   ...         ...        ...       ...        ...     ...   ...   
390   20           1          2         2          0       1     1   
391   17           2          1         0          0       0     0   
392   21           1          1         3          0       0     0   
393   18           3          1         0          0       0     0   
394   19           1          1         0          0       0     0   

     activities  nursery  higher  ...  Fjob_services  Fjob_teacher  \
0             0        1       1  ...              0             1   
1             0        0       1  ...              0             0   
2             0        1       1  ...              0             0   
3             1        1       1  ...              1             0   
4             0        1       1  ...              0             0   
..          ...      ...     ...  ...            ...           ...   
390           0        1       1  ...              1             0   
391           0        0       1  ...              1             0   
392           0        0       1  ...              0             0   
393           0        0       1  ...              0             0   
394           0        1       1  ...              0             0   

     reason_course  reason_home  reason_other  reason_reputation  \
0                1            0             0                  0   
1                1            0             0                  0   
2                0            0             1                  0   
3                0            1             0                  0   
4                0            1             0                  0   
..             ...          ...           ...                ...   
390              1            0             0                  0   
391              1            0             0                  0   
392              1            0             0                  0   
393              1            0             0                  0   
394              1            0             0                  0   

     guardian_father  guardian_mother  guardian_other  Pedu  
0                  0                1               0     8  
1                  1                0               0     2  
2                  0                1               0     2  
3                  0                1               0     6  
4                  1                0               0     6  
..               ...              ...             ...   ...  
390                0                0               1     4  
391                0                1               0     4  
392                0                0               1     2  
393                0                1               0     5  
394                1                0               0     2  

[395 rows x 48 columns]

#### Building the model and evaluating the performance

In [11]:
# spliiting the dataset into features and target
#taking 'passed' as target(y) and all other as X
X=Student.drop(['passed'],axis=1)
y=Student['passed']

In [12]:
# splitting into test and training set
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=42,test_size=0.20)

In [13]:
# checking how many data points are there in the training set
X_train.shape

(316, 49)

316 datapoints are given for training and 49 data points for testing

In [14]:
print('Grad rate of train set:{:.2f}%'.format(100 * (y_train == 1).mean()))
print('Grad rate of test set: {:.2f}%'.format(100 * (y_test == 1).mean()))

Grad rate of train set:67.41%
Grad rate of test set: 65.82%


In [15]:
# importing the required libraries
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
import time
from sklearn.metrics import f1_score

In [16]:
print ("Training using LogisticRegression with a training set size of ", len(X_train))
logit_model=LogisticRegression(random_state=42,max_iter=10000)
start =time.time()
logit_model.fit(X_train,y_train)
end = time.time()
print ("Trained model in {:.4f} seconds".format(end - start))
start=time.time()
y_pred=logit_model.predict(X_test)
end=time.time()
print ("Made predictions in {:.4f} seconds".format(end - start))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Training using LogisticRegression with a training set size of  316
Trained model in 0.0717 seconds
Made predictions in 0.0020 seconds
f1 score is: 0.825


Other than computation timing ,F1 score is used to measure the performance of the model

F1 score is the harmonic mean of Precision and Recall and gives a better measure than the Accuracy Metric.

In [21]:
print ("Training using SVM model with a training set size of ", len(X_train))
svm_model=SVC(random_state=42)
start =time.time()
svm_model.fit(X_train,y_train)
end = time.time()
print ("Trained model in {:.4f} seconds".format(end - start))
start =time.time()
y_pred=svm_model.predict(X_test)
end=time.time()
print ("Made predictions in {:.4f} seconds".format(end - start))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Training using SVM model with a training set size of  316
Trained model in 0.0120 seconds
Made predictions in 0.0050 seconds
f1 score is: 0.794


In [22]:
print ("Training using KNN with a training set size of ", len(X_train))
knn_model=KNeighborsClassifier()
start =time.time()
knn_model.fit(X_train,y_train)
end = time.time()
print ("Trained model in {:.4f} seconds".format(end - start))
start =time.time()
y_pred=knn_model.predict(X_test)
end=time.time()
print ("Made predictions in {:.4f} seconds".format(end - start))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Training using KNN with a training set size of  316
Trained model in 0.0040 seconds
Made predictions in 0.0100 seconds
f1 score is: 0.773


In [23]:
print ("Training using Random Forest with a training set size of ", len(X_train))
random_forest=RandomForestClassifier(random_state=42)
start =time.time()
random_forest.fit(X_train,y_train)
end = time.time()
print ("Trained model in {:.4f} seconds".format(end - start))
# predicing the values using X_test by the model created
start =time.time()
y_pred=random_forest.predict(X_test)
end=time.time()
print ("Made predictions in {:.4f} seconds".format(end - start))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Training using Random Forest with a training set size of  316
Trained model in 0.1551 seconds
Made predictions in 0.0313 seconds
f1 score is: 0.803


In [24]:
print ("Training using Decision tree with a training set size of ", len(X_train))
decision_tree=DecisionTreeClassifier(random_state=42)
start =time.time()
decision_tree.fit(X_train,y_train)
end = time.time()
print ("Trained model in {:.4f} seconds".format(end - start))
start =time.time()
y_pred=decision_tree.predict(X_test)
end=time.time()
print ("Made predictions in {:.4f} seconds".format(end - start))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Training using Decision tree with a training set size of  316
Trained model in 0.0090 seconds
Made predictions in 0.0030 seconds
f1 score is: 0.727


| Model | Training time | Prediciton time | F1_score
| --- | --- | --- |----|
| Logistic regression | 0.0717 | 0.0020 | 0.825
| SVM | 0.0120 | 0.0050 |  0.794
| KNN | 0.0040 | 0.0100 |  0.773
| Random Forest |0.1551 | 0.0313 |  0.803
| Decision tree | 0.0090 | 0.0030 |  0.727

####  Selecting the model

Based on the computation cost(including training time and prediction time) and F1_score, SVM performs better than other models.Since the data set is comparatively small ,SVM is the better option.